functional-samples/sample.dnr-rule-manager/manager.js

```diff
@@ -0,0 +1,199 @@
+const viewRuleButton = document.getElementById('viewRuleButton');
+
+async function getNextRuleId() {
+  const rules = await chrome.declarativeNetRequest.getDynamicRules();
+  return Math.max(...rules.map((rule) => rule.id)) + 1;
+}
+
+async function refresh() {
+  const rules = await chrome.declarativeNetRequest.getDynamicRules();
+  renderRules(rules);
+}
+
+refresh();
+
+addRuleButton.addEventListener('click', () => {
+  appendRuleItem(rulesList, {
+    id: 'NEW',
+    conditionType: 'urlFilter',
+    conditionValue: '',
+    caseSensitive: true
+  });
+});
+
+viewRuleButton.addEventListener('click', async () => {
+  const rules = await chrome.declarativeNetRequest.getDynamicRules();
+  const rulesString = JSON.stringify(rules, null, 2);
+  const newWindow = window.open();
+  newWindow.document.write(
+    `<pre style="font-size: 1rem;">${rulesString}</pre>`
+  );
+});
+
+function appendRuleItem(
+  container,
+  { id, conditionType, conditionValue, caseSensitive }
+) {
+  // Create a new rule item from the template
+  const ruleItem = ruleItemTemplate.content.cloneNode(true).children[0];
+  container.appendChild(ruleItem);
+
+  // Set the rule ID
+  ruleItem.querySelector('.rule-id').value = id;
+
+  // Set the conditionType select
+  ruleItem.querySelector('.condition-type').value = conditionType;
+
+  // set the caseSensitive input
+  ruleItem.querySelector('.case-sensitive').checked = caseSensitive;
+
+  // Set the conditionValue input
+  ruleItem.querySelector('.condition-value').value = conditionValue;
+
+  // Set button click handlers
+  function getCurrentRuleValues() {
+    const _conditionValue = ruleItem.querySelector('.condition-value').value;
+    const _conditionType = ruleItem.querySelector('.condition-type').value;
+    const _caseSensitive = ruleItem.querySelector('.case-sensitive').checked;
+    return {
+      conditionValue: _conditionValue.trim(),
+      conditionType: _conditionType,
+      caseSensitive: _caseSensitive
+    };
+  }
+  ruleItem.querySelector('.remove-rule').addEventListener('click', async () => {
+    // For a new rule, just remove the item from the DOM
+    if (id !== 'NEW') {
+      await removeRule(id);
+    }
+    ruleItem.remove();
+  });
+  ruleItem.querySelector('.save-rule').addEventListener('click', async () => {
+    const {
+      conditionValue: _conditionValue,
+      conditionType: _conditionType,
+      caseSensitive: _caseSensitive
+    } = getCurrentRuleValues();
+    const _id = id === 'NEW' ? await getNextRuleId() : id;
+    await saveRule({
+      id: _id,
+      conditionType: _conditionType,
+      conditionValue: _conditionValue,
+      caseSensitive: _caseSensitive
+    });
+    // Update the rule ID
+    ruleItem.querySelector('.rule-id').value = _id;
+    // Disable the save button
+    setSaveButtonDisabled();
+    // Update parameter
+    id = _id;
+    conditionType = _conditionType;
+    conditionValue = _conditionValue;
+    caseSensitive = _caseSensitive;
```

Comment:
Do these lines do anything?

Reply:
Perhaps I can rename these variables starting with an underscore, for example changing `_conditionValue` to `currentConditionValue`.

Reply:
I think there's too much nesting/use of parameters as variables here which is confusing.

How about having:

```
function appendRuleItem(container, initialValues) {
```

Then, rather than updating `initialValues` when something changes, we can just use `getCurrentRuleValues` whenever we need the values and also update it to return the ID if needed. Does that make sense?